# Déployez un modèle dans le cloud


# Sommaire :

**1. Préambule**<br />
&emsp;1.1 Problématique<br />
&emsp;1.2 Objectifs dans ce projet<br />
&emsp;1.3 Déroulement des étapes du projet<br />
**2. Choix techniques généraux retenus**<br />
&emsp;2.1 Calcul distribué<br />
&emsp;2.2 Transfert Learning<br />
**3. Déploiement de la solution en local**<br />
&emsp;3.1 Environnement de travail<br />
&emsp;3.2 Installation de Spark<br />
&emsp;3.3 Installation des packages<br />
&emsp;3.4 Import des librairies<br />
&emsp;3.5 Définition des PATH pour charger les images et enregistrer les résultats<br />
&emsp;3.6 Création de la SparkSession<br />
&emsp;3.7 Traitement des données<br />
&emsp;&emsp;3.7.1 Chargement des données<br />
&emsp;&emsp;3.7.2 Préparation du modèle<br />
&emsp;&emsp;3.7.3 Définition du processus de chargement des images et application <br />
&emsp;&emsp;&emsp;&emsp;&emsp;de leur featurisation à travers l'utilisation de pandas UDF<br />
&emsp;&emsp;3.7.4 Exécution des actions d'extractions de features<br />
&emsp;3.8 Chargement des données enregistrées et validation du résultat<br />
**4. Déploiement de la solution sur le cloud**<br />
&emsp;4.1 Choix du prestataire cloud : AWS<br />
&emsp;4.2 Choix de la solution technique : EMR<br />
&emsp;4.3 Choix de la solution de stockage des données : Amazon S3<br />
&emsp;4.4 Configuration de l'environnement de travail<br />
&emsp;4.5 Upload de nos données sur S3<br />
&emsp;4.6 Configuration du serveur EMR<br />
&emsp;&emsp;4.6.1 Étape 1 : Logiciels et étapes<br />
&emsp;&emsp;&emsp;4.6.1.1 Configuration des logiciels<br />
&emsp;&emsp;&emsp;4.6.1.2 Modifier les paramètres du logiciel<br />
&emsp;&emsp;4.6.2 Étape 2 : Matériel<br />
&emsp;&emsp;4.6.3 Étape 3 : Paramètres de cluster généraux<br />
&emsp;&emsp;&emsp;4.6.3.1 Options générales<br />
&emsp;&emsp;&emsp;4.6.3.2 Actions d'amorçage<br />
&emsp;&emsp;4.6.4 Étape 4 : Sécurité<br />
&emsp;&emsp;&emsp;4.6.4.1 Options de sécurité<br />
&emsp;4.7 Instanciation du serveur<br />
&emsp;4.8 Création du tunnel SSH à l'instance EC2 (Maître)<br />
&emsp;&emsp;4.8.1 Création des autorisations sur les connexions entrantes<br />
&emsp;&emsp;4.8.2 Création du tunnel ssh vers le Driver<br />
&emsp;&emsp;4.8.3 Configuration de FoxyProxy<br />
&emsp;&emsp;4.8.4 Accès aux applications du serveur EMR via le tunnel ssh<br />
&emsp;4.9 Connexion au notebook JupyterHub<br />
&emsp;4.10 Exécution du code<br />
&emsp;&emsp;4.10.1 Démarrage de la session Spark<br />
&emsp;&emsp;4.10.2 Installation des packages<br />
&emsp;&emsp;4.10.3 Import des librairies<br />
&emsp;&emsp;4.10.4 Définition des PATH pour charger les images et enregistrer les résultats<br />
&emsp;&emsp;4.10.5 Traitement des données<br />
&emsp;&emsp;&emsp;4.10.5.1 Chargement des données<br />
&emsp;&emsp;&emsp;4.10.5.2 Préparation du modèle<br />
&emsp;&emsp;&emsp;4.10.5.3 Définition du processus de chargement des images<br />
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;et application de leur featurisation à travers l'utilisation de pandas UDF<br />
&emsp;&emsp;&emsp;4.10.5.4 Exécutions des actions d'extractions de features<br />
&emsp;&emsp;4.10.6 Chargement des données enregistrées et validation du résultat<br />
&emsp;4.11 Suivi de l'avancement des tâches avec le Serveur d'Historique Spark<br />
&emsp;4.12 Résiliation de l'instance EMR<br />
&emsp;4.13 Cloner le serveur EMR (si besoin)<br />
&emsp;4.14 Arborescence du serveur S3 à la fin du projet<br />
**5. Conclusion**

# 1. Préambule

## 1.1 Problématique

La très jeune start-up de l'AgriTech, nommée "**Fruits**!", <br />
cherche à proposer des solutions innovantes pour la récolte des fruits.

La volonté de l’entreprise est de préserver la biodiversité des fruits <br />
en permettant des traitements spécifiques pour chaque espèce de fruits <br />
en développant des robots cueilleurs intelligents.

La start-up souhaite dans un premier temps se faire connaître en mettant <br />
à disposition du grand public une application mobile qui permettrait aux <br />
utilisateurs de prendre en photo un fruit et d'obtenir des informations sur ce fruit.

Pour la start-up, cette application permettrait de sensibiliser le grand public <br />
à la biodiversité des fruits et de mettre en place une première version du moteur <br />
de classification des images de fruits.

De plus, le développement de l’application mobile permettra de construire <br />
une première version de l'architecture **Big Data** nécessaire.

## 1.2 Objectifs dans ce projet

1. Développer une première chaîne de traitement des données qui <br />
   comprendra le **preprocessing** et une étape de **réduction de dimension**.
2. Tenir compte du fait que <u>le volume de données va augmenter <br />
   très rapidement</u> après la livraison de ce projet, ce qui implique de:
 - Déployer le traitement des données dans un environnement **Big Data**
 - Développer les scripts en **pyspark** pour effectuer du **calcul distribué**

## 1.3 Déroulement des étapes du projet

Le projet va être réalisé en 2 temps, dans deux environnements différents. <br />
Nous allons dans un premier temps développer et exécuter notre code en local, <br />
en travaillant sur un nombre limité d'images à traiter.

Une fois les choix techniques validés, nous déploierons notre solution <br />
dans un environnement Big Data en mode distribué.

<u>Pour cette raison, ce projet sera divisé en 3 parties</u>:
1. Liste des choix techniques généraux retenus
2. Déploiement de la solution en local
3. Déploiement de la solution dans le cloud

# 2. Choix techniques généraux retenus

## 2.1 Calcul distribué

L’énoncé du projet nous impose de développer des scripts en **pyspark** <br />
afin de <u>prendre en compte l’augmentation très rapide du volume <br />
de donné après la livraison du projet</u>.

Pour comprendre rapidement et simplement ce qu’est **pyspark** <br />
et son principe de fonctionnement, nous vous conseillons de lire <br />
cet article : [PySpark : Tout savoir sur la librairie Python](https://datascientest.com/pyspark)

<u>Le début de l’article nous dit ceci </u>:<br />
« *Lorsque l’on parle de traitement de bases de données sur python, <br />
on pense immédiatement à la librairie pandas. Cependant, lorsqu’on a <br />
affaire à des bases de données trop massives, les calculs deviennent trop lents.<br />
Heureusement, il existe une autre librairie python, assez proche <br />
de pandas, qui permet de traiter des très grandes quantités de données : PySpark.<br />
Apache Spark est un framework open-source développé par l’AMPLab <br />
de UC Berkeley permettant de traiter des bases de données massives <br />
en utilisant le calcul distribué, technique qui consiste à exploiter <br />
plusieurs unités de calcul réparties en clusters au profit d’un seul <br />
projet afin de diviser le temps d’exécution d’une requête.<br />
Spark a été développé en Scala et est au meilleur de ses capacités <br />
dans son langage natif. Cependant, la librairie PySpark propose de <br />
l’utiliser avec le langage Python, en gardant des performances <br />
similaires à des implémentations en Scala.<br />
Pyspark est donc une bonne alternative à la librairie pandas lorsqu’on <br />
cherche à traiter des jeux de données trop volumineux qui entraînent <br />
des calculs trop chronophages.* »

Comme nous le constatons, **pySpark** est un moyen de communiquer <br />
avec **Spark** via le langage **Python**.<br />
**Spark**, quant à lui, est un outil qui permet de gérer et de coordonner <br />
l'exécution de tâches sur des données à travers un groupe d'ordinateurs. <br />
<u>Spark (ou Apache Spark) est un framework open source de calcul distribué <br />
in-memory pour le traitement et l'analyse de données massives</u>.

Un autre [article très intéressant et beaucoup plus complet pour <br />
comprendre le **fonctionnement de Spark**](https://www.veonum.com/apache-spark-pour-les-nuls/), ainsi que le rôle <br />
des **Spark Session** que nous utiliserons dans ce projet.

<u>Voici également un extrait</u>:

*Les applications Spark se composent d’un pilote (« driver process ») <br />
et de plusieurs exécuteurs (« executor processes »). Il peut être configuré <br />
pour être lui-même l’exécuteur (local mode) ou en utiliser autant que <br />
nécessaire pour traiter l’application, Spark prenant en charge la mise <br />
à l’échelle automatique par une configuration d’un nombre minimum <br />
et maximum d’exécuteurs.*

![Schéma de Spark](img/spark-schema.png)

*Le driver (parfois appelé « Spark Session ») distribue et planifie <br />
les tâches entre les différents exécuteurs qui les exécutent et permettent <br />
un traitement réparti. Il est le responsable de l’exécution du code <br />
sur les différentes machines.

Chaque exécuteur est un processus Java Virtual Machine (JVM) distinct <br />
dont il est possible de configurer le nombre de CPU et la quantité de <br />
mémoire qui lui est alloué. <br />
Une seule tâche peut traiter un fractionnement de données à la fois.*

Dans les deux environnements (Local et Cloud) nous utiliserons donc **Spark** <br />
et nous l’exploiterons à travers des scripts python grâce à **PySpark**.

Dans la <u>version locale</u> de notre script nous **simulerons <br />
le calcul distribué** afin de valider que notre solution fonctionne.<br />
Dans la <u>version cloud</u> nous **réaliserons les opérations sur un cluster de machine**.

## 2.2 Transfert Learning

L'énoncé du projet nous demande également de <br />
réaliser une première chaîne de traitement <br />
des données qui comprendra le preprocessing et <br />
une étape de réduction de dimension.

Il est également précisé qu'il n'est pas nécessaire <br />
d'entraîner un modèle pour le moment.

Nous décidons de partir sur une solution de **transfert learning**.

Simplement, le **transfert learning** consiste <br />
à utiliser la connaissance déjà acquise <br />
par un modèle entraîné (ici **MobileNetV2**) pour <br />
l'adapter à notre problématique.

Nous allons fournir au modèle nos images, et nous allons <br />
<u>récupérer l'avant dernière couche</u> du modèle.<br />
En effet la dernière couche de modèle est une couche softmax <br />
qui permet la classification des images ce que nous ne <br />
souhaitons pas dans ce projet.

L'avant dernière couche correspond à un **vecteur <br />
réduit** de dimension (1,1,1280).

Cela permettra de réaliser une première version du moteur <br />
pour la classification des images des fruits.

**MobileNetV2** a été retenu pour sa <u>rapidité d'exécution</u>, <br />
particulièrement adaptée pour le traitement d'un gros volume <br />
de données ainsi que la <u>faible dimensionnalité du vecteur <br />
de caractéristique en sortie</u> (1,1,1280)

# 3. Déploiement de la solution en local


## 3.1 Environnement de travail

Pour des raisons de simplicité, nous développons dans un environnement <br />
Linux Unbuntu (exécuté depuis une machine Windows dans une machine virtuelle)
* Pour installer une machine virtuelle :  https://www.malekal.com/meilleurs-logiciels-de-machine-virtuelle-gratuits-ou-payants/

## 3.2 Installation de Spark

[La première étape consiste à installer Spark ](https://computingforgeeks.com/how-to-install-apache-spark-on-ubuntu-debian/)

## 3.3 Installation des packages

<u>On installe ensuite à l'aide de la commande **pip** <br />
les packages qui nous seront nécessaires</u> :

In [1]:
!pip install Pandas pillow tensorflow pyspark pyarrow --quiet

## 3.4 Import des librairies

In [2]:
import pandas as pd
from PIL import Image
import numpy as np
import io
import os

import tensorflow as tf
from tensorflow.keras.applications.mobilenet_v2 import MobileNetV2, preprocess_input
from tensorflow.keras.preprocessing.image import img_to_array
from tensorflow.keras import Model
from pyspark.sql.functions import col, pandas_udf, PandasUDFType, element_at, split, udf
from pyspark.sql import SparkSession
from pyspark.ml.feature import PCA
from pyspark.ml.linalg import Vectors, VectorUDT

## 3.5 Définition des PATH pour charger les images <br /> et enregistrer les résultats

Dans cette version locale nous partons du principe que les données <br />
sont stockées dans le même répertoire que le notebook.<br />
Nous utilisons les **103 images** du dossier test-multiple_fruits pour valider notre pipeline en local.<br />
L'extrait des images à charger est stockée dans le dossier **Test1**.<br />
Nous enregistrerons le résultat de notre traitement <br />
dans le dossier "**Results_Local**"

In [3]:
# 1. Montage Google Drive
from google.colab import drive
drive.mount('/content/drive')
import os
import zipfile

# 2. Navigation vers le projet
drive_folder = "/content/drive/MyDrive/Colab_Notebooks/Project_11/"
os.chdir(drive_folder)

# 3. Variables PATH adaptées pour test-multiple-fruits
PATH = os.getcwd()
fruits_folder = 'test-multiple_fruits'
PATH_Data = PATH + f'/data/{fruits_folder}'
PATH_Result = PATH + '/data/Results'

# 4. Vérification si déjà extrait
all_extracted = False
if os.path.exists(PATH_Data):
    existing_files = [f for f in os.listdir(PATH_Data) if f.endswith('.jpg')]
    if len(existing_files) > 0:
        all_extracted = True

# 5. Extraction seulement si nécessaire
if not all_extracted:
    print("Extraction des données test-multiple-fruits en cours...")

    # Nettoyer d'abord le dossier existant
    if os.path.exists(PATH_Data):
        import shutil
        shutil.rmtree(PATH_Data)

    os.makedirs('data', exist_ok=True)
    os.makedirs(PATH_Data, exist_ok=True)

    with zipfile.ZipFile('fruits.zip', 'r') as zip_ref:
        for file in zip_ref.namelist():
            if f'fruits-360_dataset/fruits-360/{fruits_folder}/' in file and file.endswith('.jpg'):
                zip_ref.extract(file, '.')
                filename = file.split('/')[-1]
                new_path = f'data/{fruits_folder}/{filename}'
                os.rename(file, new_path)

    # Nettoyer
    import shutil
    if os.path.exists('fruits-360_dataset'):
        shutil.rmtree('fruits-360_dataset')

    print("Extraction terminée.")
else:
    print("Données déjà extraites, pas besoin de re-extraire.")

# 6. Vérification finale
print(f"PATH: {PATH}")
print(f"PATH_Data: {PATH_Data}")
print(f"PATH_Result: {PATH_Result}")
if os.path.exists(PATH_Data):
    image_files = [f for f in os.listdir(PATH_Data) if f.endswith('.jpg')]
    print(f"Images disponibles: {len(image_files)} fichiers")
    if len(image_files) > 0:
        print(f"Exemples: {image_files[:5]}")

Mounted at /content/drive
Données déjà extraites, pas besoin de re-extraire.
PATH: /content/drive/MyDrive/Colab_Notebooks/Project_11
PATH_Data: /content/drive/MyDrive/Colab_Notebooks/Project_11/data/test-multiple_fruits
PATH_Result: /content/drive/MyDrive/Colab_Notebooks/Project_11/data/Results
Images disponibles: 103 fichiers
Exemples: ['Bananas(lady_finger)1.jpg', 'Bananas(lady_finger)2.jpg', 'Bananas(lady_finger)3.jpg', 'Bananas(lady_finger)4.jpg', 'Bananas(lady_finger)5.jpg']


## 3.6 Création de la SparkSession

L’application Spark est contrôlée grâce à un processus de pilotage (driver process) appelé **SparkSession**. <br />
<u>Une instance de **SparkSession** est la façon dont Spark exécute les fonctions définies par l’utilisateur <br />
dans l’ensemble du cluster</u>. <u>Une SparkSession correspond toujours à une application Spark</u>.

<u>Ici nous créons une session spark en spécifiant dans l'ordre</u> :
 1. un **nom pour l'application**, qui sera affichée dans l'interface utilisateur Web Spark "**P11**"
 2. que l'application doit s'exécuter **localement**. <br />
   Nous ne définissons pas le nombre de cœurs à utiliser (comme .master('local[4]) pour 4 cœurs à utiliser), <br />
   nous utiliserons donc tous les cœurs disponibles dans notre processeur.<br />
 3. une option de configuration supplémentaire permettant d'utiliser le **format "parquet"** <br />
   que nous utiliserons pour enregistrer et charger le résultat de notre travail.
 4. vouloir **obtenir une session spark** existante ou si aucune n'existe, en créer une nouvelle

In [4]:
spark = (SparkSession
             .builder
             .appName('P11')
             .master('local')
             .config("spark.sql.parquet.writeLegacyFormat", 'true')
             .getOrCreate()
)

<u>Nous créons également la variable "**sc**" qui est un **SparkContext** issue de la variable **spark**</u> :

In [5]:
sc = spark.sparkContext

<u>Affichage des informations de Spark en cours d'execution</u> :

In [6]:
spark

## 3.7 Traitement des données

<u>Dans la suite de notre flux de travail, <br />
nous allons successivement</u> :
1. Préparer nos données
    1. Importer les images dans un dataframe **pandas UDF**
    2. Associer aux images leur **label**
    3. Préprocesser en **redimensionnant nos images pour <br />
       qu'elles soient compatibles avec notre modèle**
2. Préparer notre modèle
    1. Importer le modèle **MobileNetV2**
    2. Créer un **nouveau modèle** dépourvu de la dernière couche de MobileNetV2
3. Définir le processus de chargement des images et l'application <br />
   de leur featurisation à travers l'utilisation de pandas UDF
3. Exécuter les actions d'extraction de features
4. Enregistrer le résultat de nos actions
5. Tester le bon fonctionnement en chargeant les données enregistrées




### 3.7.1 Chargement des données

Les images sont chargées au format binaire, ce qui offre, <br />
plus de souplesse dans la façon de prétraiter les images.

Avant de charger les images, nous spécifions que nous voulons charger <br />
uniquement les fichiers dont l'extension est **jpg**.

Nous indiquons également de charger tous les objets possibles contenus <br />
dans les sous-dossiers du dossier communiqué.

In [7]:
images = spark.read.format("binaryFile") \
  .option("pathGlobFilter", "*.jpg") \
  .option("recursiveFileLookup", "true") \
  .load(PATH_Data)

<u>Affichage des 5 premières images contenant</u> :
 - le path de l'image
 - la date et heure de sa dernière modification
 - sa longueur
 - son contenu encodé en valeur hexadécimal

<u>Je ne conserve que le **path** de l'image et j'ajoute <br />
    une colonne contenant les **labels** de chaque image</u> :

In [8]:
images = images.withColumn('label', element_at(split(images['path'], '/'),-2))
print(images.printSchema())
print(images.select('path','label').show(5,False))

root
 |-- path: string (nullable = true)
 |-- modificationTime: timestamp (nullable = true)
 |-- length: long (nullable = true)
 |-- content: binary (nullable = true)
 |-- label: string (nullable = true)

None
+--------------------------------------------------------------------------------------------------+--------------------+
|path                                                                                              |label               |
+--------------------------------------------------------------------------------------------------+--------------------+
|file:/content/drive/MyDrive/Colab_Notebooks/Project_11/data/test-multiple_fruits/cherries.jpg     |test-multiple_fruits|
|file:/content/drive/MyDrive/Colab_Notebooks/Project_11/data/test-multiple_fruits/strawberries.jpg |test-multiple_fruits|
|file:/content/drive/MyDrive/Colab_Notebooks/Project_11/data/test-multiple_fruits/apples4.jpg      |test-multiple_fruits|
|file:/content/drive/MyDrive/Colab_Notebooks/Project_11/da

### 3.7.2 Préparation du modèle

Je vais utiliser la technique du **transfert learning** pour extraire les features des images.<br />
J'ai choisi d'utiliser le modèle **MobileNetV2** pour sa rapidité d'exécution comparée <br />
à d'autres modèles comme *VGG16* par exemple.

Pour en savoir plus sur la conception et le fonctionnement de MobileNetV2, <br />
je vous invite à lire [cet article](https://towardsdatascience.com/review-mobilenetv2-light-weight-model-image-classification-8febb490e61c).

<u>Voici le schéma de son architecture globale</u> :

![Architecture de MobileNetV2](img/mobilenetv2_architecture.png)

Il existe une dernière couche qui sert à classer les images <br />
selon 1000 catégories que nous ne voulons pas utiliser.<br />
L'idée dans ce projet est de récupérer le **vecteur de caractéristiques <br />
de dimensions (1,1,1280)** qui servira, plus tard, au travers d'un moteur <br />
de classification à reconnaitre les différents fruits du jeu de données.

Comme d'autres modèles similaires, **MobileNetV2**, lorsqu'on l'utilise <br />
en incluant toutes ses couches, attend obligatoirement des images <br />
de dimension (224,224,3). Nos images étant toutes de dimension (100,100,3), <br />
nous devrons simplement les **redimensionner** avant de les confier au modèle.

<u>Dans l'odre</u> :
 1. Nous chargeons le modèle **MobileNetV2** avec les poids **précalculés** <br />
    issus d'**imagenet** et en spécifiant le format de nos images en entrée
 2. Nous créons un nouveau modèle avec:
  - <u>en entrée</u> : l'entrée du modèle MobileNetV2
  - <u>en sortie</u> : l'avant dernière couche du modèle MobileNetV2

In [9]:
model = MobileNetV2(weights='imagenet',
                    include_top=True,
                    input_shape=(224, 224, 3))

14536120/14536120 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [10]:
new_model = Model(inputs=model.input,
                  outputs=model.layers[-2].output)

Affichage du résumé de notre nouveau modèle où nous constatons <br />
que <u>nous récupérons bien en sortie un vecteur de dimension (1, 1, 1280)</u> :

In [11]:
new_model.summary()

Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer         │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1 (Conv2D)      │ (None, 112, 112,  │        864 │ input_layer[0][0] │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ bn_Conv1            │ (None, 112, 112,  │        128 │ Conv1[0][0]       │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ Conv1_relu (ReLU)   │ (None, 112, 112,  │          0 │ bn_Conv1[0][0]    │
│                     │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        288 │ Conv1_relu[0][0]  │
│ (DepthwiseConv2D)   │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │        128 │ expanded_conv_de… │
│ (BatchNormalizatio… │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_dept… │ (None, 112, 112,  │          0 │ expanded_conv_de… │
│ (ReLU)              │ 32)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │        512 │ expanded_conv_de… │
│ (Conv2D)            │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ expanded_conv_proj… │ (None, 112, 112,  │         64 │ expanded_conv_pr… │
│ (BatchNormalizatio… │ 16)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand      │ (None, 112, 112,  │      1,536 │ expanded_conv_pr… │
│ (Conv2D)            │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_BN   │ (None, 112, 112,  │        384 │ block_1_expand[0… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_expand_relu │ (None, 112, 112,  │          0 │ block_1_expand_B… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_pad         │ (None, 113, 113,  │          0 │ block_1_expand_r… │
│ (ZeroPadding2D)     │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise   │ (None, 56, 56,    │        864 │ block_1_pad[0][0] │
│ (DepthwiseConv2D)   │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │        384 │ block_1_depthwis… │
│ (BatchNormalizatio… │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_depthwise_… │ (None, 56, 56,    │          0 │ block_1_depthwis… │
│ (ReLU)              │ 96)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ block_1_project     │ (None, 56, 56,    │      2,304 │ block_1_depthwis

 Total params: 2,257,984 (8.61 MB)

 Trainable params: 2,223,872 (8.48 MB)

 Non-trainable params: 34,112 (133.25 KB)

Tous les workeurs doivent pouvoir accéder au modèle ainsi qu'à ses poids. <br />
Une bonne pratique consiste à charger le modèle sur le driver puis à diffuser <br />
ensuite les poids aux différents workeurs.

In [12]:
brodcast_weights = sc.broadcast(new_model.get_weights())

<u>Mettons cela sous forme de fonction</u> :

In [13]:
def model_fn():
    """
    Returns a MobileNetV2 model with top layer removed
    and broadcasted pretrained weights.
    """
    model = MobileNetV2(weights=None, # There is no need to download pre-trained imagenet weights, as we already have broadcast_weights.
                        include_top=True,
                        input_shape=(224, 224, 3))
    for layer in model.layers:
        layer.trainable = False
    new_model = Model(inputs=model.input,
                  outputs=model.layers[-2].output)
    new_model.set_weights(brodcast_weights.value)
    return new_model

### 3.7.3 Définition du processus de chargement des images et application <br/>de leur featurisation à travers l'utilisation de pandas UDF

Ce notebook définit la logique par étapes, jusqu'à Pandas UDF.

<u>L'empilement des appels est la suivante</u> :

- Pandas UDF
  - featuriser une série d'images pd.Series
   - prétraiter une image

In [14]:
def preprocess(content):
    """
    Preprocesses raw image bytes for prediction.
    """
    img = Image.open(io.BytesIO(content)).resize([224, 224])
    arr = img_to_array(img)
    return preprocess_input(arr)

def featurize_series(model, content_series):
    """
    Featurize a pd.Series of raw images using the input model.
    :return: a pd.Series of image features
    """
    input = np.stack(content_series.map(preprocess))
    preds = model.predict(input)
    # For some layers, output features will be multi-dimensional tensors.
    # We flatten the feature tensors to vectors for easier storage in Spark DataFrames.
    output = [p.flatten() for p in preds]
    return pd.Series(output)

@pandas_udf('array<float>', PandasUDFType.SCALAR_ITER)
def featurize_udf(content_series_iter):
    '''
    This method is a Scalar Iterator pandas UDF wrapping our featurization function.
    The decorator specifies that this returns a Spark DataFrame column of type ArrayType(FloatType).

    :param content_series_iter: This argument is an iterator over batches of data, where each batch
                              is a pandas Series of image data.
    '''
    # With Scalar Iterator pandas UDFs, we can load the model once and then re-use it
    # for multiple data batches.  This amortizes the overhead of loading big models.
    model = model_fn()
    for content_series in content_series_iter:
        yield featurize_series(model, content_series)

/usr/local/lib/python3.12/dist-packages/pyspark/sql/pandas/functions.py:407: UserWarning: In Python 3.6+ and Spark 3.0+, it is preferred to specify type hints for pandas UDF instead of specifying pandas UDF type which will be deprecated in the future releases. See SPARK-28264 for more details.
  warnings.warn(


### 3.7.4 Exécution des actions d'extraction de features

Les Pandas UDF, sur de grands enregistrements (par exemple, de très grandes images), <br />
peuvent rencontrer des erreurs de type Out Of Memory (OOM).<br />
Si vous rencontrez de telles erreurs dans la cellule ci-dessous, <br />
essayez de réduire la taille du lot Arrow via 'maxRecordsPerBatch'

Je n'utiliserai pas cette commande dans ce projet <br />
et je laisse donc la commande en commentaire.

In [15]:
# spark.conf.set("spark.sql.execution.arrow.maxRecordsPerBatch", "1024")

Nous pouvons maintenant exécuter la featurisation sur l'ensemble de notre DataFrame Spark.<br />
<u>REMARQUE</u> : Cela peut prendre beaucoup de temps, tout dépend du volume de données à traiter. <br />

Notre jeu de données **test-multiple_fruits** contient environ **103 images**. <br />
Dans l'exécution en mode **local**, <br />
nous <u>traiterons cet ensemble complet d'images multi-fruits</u>, parfait pour valider notre pipeline avant le déploiement EMR.

In [16]:
features_df = images.repartition(20).select(col("path"),
                                            col("label"),
                                            featurize_udf("content").alias("features")
                                           )

<u>Rappel du PATH où seront inscrits les fichiers au format "**parquet**" <br />
contenant nos résultats, à savoir, un DataFrame contenant 3 colonnes</u> :
 1. Path des images
 2. Label de l'image
 3. Vecteur de caractéristiques de l'image

In [17]:
print(PATH_Result)

/content/drive/MyDrive/Colab_Notebooks/Project_11/data/Results


<u>Enregistrement des données traitées au format "**parquet**"</u> :

In [18]:
features_df.write.mode("overwrite").parquet(PATH_Result)

### 3.7.5 Exécution d'une PCA distribuée

#### Objectif de cette étape

L'extraction de features via MobileNetV2 génère des vecteurs de **1280 dimensions** par image. Pour un passage à l'échelle industrielle, cette haute dimensionnalité pose plusieurs défis :

- **Stockage** : 1280 dimensions × millions d'images = volumes de données massifs
- **Performance** : Les algorithmes de classification sont plus lents sur des données haute dimension
- **Mémoire** : Contraintes lors du chargement en mémoire pour l'entraînement de modèles

#### Pourquoi une PCA distribuée ?

La PCA (Principal Component Analysis) réduit la dimensionnalité en conservant les **composantes principales** qui capturent le maximum de variance des données. Dans notre contexte Big Data :

1. **Calcul distribué** : PySpark répartit le calcul de la matrice de covariance sur plusieurs machines
2. **Passage à l'échelle** : Traitement possible de millions d'images sans limitation mémoire
3. **Optimisation** : Réduction de 1280 → 70 dimensions (compression élevée) tout en conservant l'information essentielle

#### Implémentation technique

La PCA distribuée nécessite une conversion des features au format `Vector` de Spark ML, puis l'application de l'algorithme PCA sur l'ensemble du dataset de manière parallélisée.

#### Limitation

Il est important de noter que la présence d'un faible échantillon (exemple minimal) va influer sur la variance captée par chacune des composantes principales. 100 composantes principales pour un échantillon de test de 102 images va forcément mathématiquement capter presque 100% de la variance. Cela sera bien moins vrai pour un échantillon de plus de 20 000 images. La variable `k_components` sera donc probablement à ajuster, en particulier dans l'environnement cloud.

In [19]:
k_components = 70

# Convertir les features en format ML de Spark
def to_vector(features):
    return Vectors.dense(features)

vector_udf = udf(to_vector, VectorUDT())

# Préparer les données pour la PCA
features_vector_df = features_df.withColumn("features_vector", vector_udf("features"))

# Appliquer la PCA pour réduire de 1280 à k_components dimensions
pca = PCA(k=k_components, inputCol="features_vector", outputCol="pca_features")
pca_model = pca.fit(features_vector_df)
pca_result = pca_model.transform(features_vector_df)

# Afficher la variance expliquée par les composantes principales
explained_variance = pca_model.explainedVariance
total_variance = sum(explained_variance)
print(f"Variance expliquée par les {k_components} premières composantes : {total_variance:.4f} ({total_variance*100:.2f}%)")
print(f"Réduction de dimensionnalité : 1280 → {k_components} dimensions")

# Sélectionner les colonnes finales
final_df = pca_result.select("path", "label", "pca_features")

# Sauvegarder les résultats avec PCA localement
PATH_Result_PCA = PATH + '/data/Results_PCA'
final_df.write.mode("overwrite").parquet(PATH_Result_PCA)

print(f"Résultats sauvés localement : {PATH_Result_PCA}")

Variance expliquée par les 70 premières composantes : 0.9576 (95.76%)
Réduction de dimensionnalité : 1280 → 70 dimensions
Résultats sauvés localement : /content/drive/MyDrive/Colab_Notebooks/Project_11/data/Results_PCA


### 3.7.6 Comparaison des bases Parquet avec et sans PCA

Cette analyse permet de quantifier l'impact de la réduction de dimensions sur l'espace de stockage et de valider l'intérêt de la PCA dans un contexte Big Data.

Nous comparons deux versions des données :
- **Features brutes** : 1280 dimensions par image (PATH_Result)  
- **Features PCA** : 70 dimensions par image (PATH_Result_PCA)

In [20]:
def get_folder_size(folder_path):
    """Calcule la taille totale d'un dossier en octets"""
    total_size = 0
    for dirpath, dirnames, filenames in os.walk(folder_path):
        for filename in filenames:
            file_path = os.path.join(dirpath, filename)
            total_size += os.path.getsize(file_path)
    return total_size

def format_bytes(bytes_value):
    """Convertit les octets en format lisible"""
    for unit in ['B', 'KB', 'MB', 'GB']:
        if bytes_value < 1024.0:
            return f"{bytes_value:.2f} {unit}"
        bytes_value /= 1024.0
    return f"{bytes_value:.2f} TB"

# Analyse des tailles de stockage
print("=== ANALYSE COMPARATIVE DES FICHIERS PARQUET ===\n")

if os.path.exists(PATH_Result):
    size_original = get_folder_size(PATH_Result)
    print(f"📊 Features originales (1280D) : {format_bytes(size_original)}")
else:
    print("❌ Dossier PATH_Result non trouvé")

if os.path.exists(PATH_Result_PCA):
    size_pca = get_folder_size(PATH_Result_PCA)
    print(f"🔬 Features PCA (100D)         : {format_bytes(size_pca)}")
else:
    print("❌ Dossier PATH_Result_PCA non trouvé")

# Calculs de compression
if os.path.exists(PATH_Result) and os.path.exists(PATH_Result_PCA):
    reduction_ratio = (size_original - size_pca) / size_original * 100
    compression_ratio = size_original / size_pca

    print(f"\n📈 GAINS DE LA PCA :")
    print(f"   • Réduction de taille : {reduction_ratio:.1f}%")
    print(f"   • Facteur de compression : {compression_ratio:.1f}x")
    print(f"   • Espace économisé : {format_bytes(size_original - size_pca)}")

    # Projection pour un dataset complet
    print(f"\n🔮 PROJECTION POUR 22,688 IMAGES (dataset Test complet) :")
    nb_images_test = len([f for f in os.listdir(PATH_Data) if f.endswith('.jpg')])
    scale_factor = 22688 / nb_images_test

    projected_original = size_original * scale_factor
    projected_pca = size_pca * scale_factor
    projected_savings = projected_original - projected_pca

    print(f"   • Features 1280D : {format_bytes(projected_original)}")
    print(f"   • Features 100D  : {format_bytes(projected_pca)}")
    print(f"   • Économies      : {format_bytes(projected_savings)}")

=== ANALYSE COMPARATIVE DES FICHIERS PARQUET ===

📊 Features originales (1280D) : 455.99 KB
🔬 Features PCA (100D)         : 110.69 KB

📈 GAINS DE LA PCA :
   • Réduction de taille : 75.7%
   • Facteur de compression : 4.1x
   • Espace économisé : 345.30 KB

🔮 PROJECTION POUR 22,688 IMAGES (dataset Test complet) :
   • Features 1280D : 98.09 MB
   • Features 100D  : 23.81 MB
   • Économies      : 74.28 MB


## 3.8 Chargement des données enregistrées et validation du résultat

<u>On charge les données fraichement enregistrées dans un **DataFrame Pandas**</u> :

In [21]:
df_pca = pd.read_parquet(PATH_Result_PCA, engine='pyarrow')

In [22]:
PATH_Result_PCA

'/content/drive/MyDrive/Colab_Notebooks/Project_11/data/Results_PCA'

<u>On affiche les 5 premières lignes du DataFrame</u> :

In [23]:
df_pca.head()

path                 label  \
0  file:/content/drive/MyDrive/Colab_Notebooks/Pr...  test-multiple_fruits   
1  file:/content/drive/MyDrive/Colab_Notebooks/Pr...  test-multiple_fruits   
2  file:/content/drive/MyDrive/Colab_Notebooks/Pr...  test-multiple_fruits   
3  file:/content/drive/MyDrive/Colab_Notebooks/Pr...  test-multiple_fruits   
4  file:/content/drive/MyDrive/Colab_Notebooks/Pr...  test-multiple_fruits   

                                        pca_features  
0  {'type': 1, 'size': None, 'indices': None, 'va...  
1  {'type': 1, 'size': None, 'indices': None, 'va...  
2  {'type': 1, 'size': None, 'indices': None, 'va...  
3  {'type': 1, 'size': None, 'indices': None, 'va...  
4  {'type': 1, 'size': None, 'indices': None, 'va...

<u>On valide que la dimension du vecteur de caractéristiques des images est bien de dimension 1280</u> :

In [24]:
# Explorer la structure du vecteur PCA
pca_vector = df_pca.loc[0,'pca_features']
print(f"Type: {type(pca_vector)}")
print(f"Clés disponibles: {pca_vector.keys()}")
print(f"Structure complète: {pca_vector}")

# Accéder aux valeurs du vecteur
if 'values' in pca_vector:
    values = pca_vector['values']
    print(f"Nombre de dimensions PCA: {len(values)}")
    print(f"Premières valeurs: {values[:5]}")

Type: <class 'dict'>
Clés disponibles: dict_keys(['type', 'size', 'indices', 'values'])
Structure complète: {'type': 1, 'size': None, 'indices': None, 'values': array([ 1.09079279e+01, -2.58490118e+00,  7.87460909e+00,  3.81716565e+00,
       -3.27298081e+00, -4.01129233e-01,  1.96036631e+00, -4.19159324e+00,
        3.50820197e+00, -3.91018984e+00, -5.62316805e+00,  5.02529295e-01,
        5.92593268e+00,  1.80210053e+00, -3.53591062e-01, -2.79969543e+00,
       -3.43373668e+00,  4.62941873e-01,  9.85715375e-01, -7.72570108e-01,
       -2.47998529e+00,  1.41398007e+00,  2.58160073e+00,  2.38604316e-01,
        1.69389754e+00,  5.10729600e+00,  4.99079239e+00,  3.06147872e+00,
       -4.16842221e+00, -8.70391555e-01, -5.98061705e-01, -2.44537620e+00,
       -4.95345172e+00,  3.33079133e+00,  4.72878440e+00,  9.31145228e-01,
        4.31761611e-01,  3.43800270e-01,  2.11200652e+00,  3.91346604e+00,
       -1.08333814e+00, -3.14158530e+00,  1.62085391e+00,  1.74567617e+00,
        2.0178

In [25]:
df_pca.shape

(103, 3)

Nous venons de valider le processus sur le jeu de données test-multiple_fruits en local où nous avons simulé un cluster de machines en répartissant la charge de travail sur différents cœurs de processeur au sein d'une même machine.

Nous allons maintenant généraliser le processus en déployant notre solution sur un réel cluster de machines EMR.

# 4. Déploiement de la solution sur le cloud

Maintenant que nous avons vérifié que notre solution fonctionne, <br />
il est temps de la <u>déployer à plus grande échelle sur un vrai cluster de machines</u>.

**Attention**, *je travaille sous Linux avec une version Ubuntu, <br />
les commandes décrites ci-dessous sont donc réalisées <br />
exclusivement dans cet environnement.*

<u>Plusieurs contraintes se posent</u> :
 1. Quel prestataire de Cloud choisir ?
 2. Quelles solutions de ce prestataire adopter ?
 3. Où stocker nos données ?
 4. Comment configurer nos outils dans ce nouvel environnement ?

## 4.1 Choix du prestataire cloud : AWS

Le prestataire le plus connu et qui offre à ce jour l'offre <br />
la plus large dans le cloud computing est **Amazon Web Services** (AWS).<br />
Certaines de leurs offres sont parfaitement adaptées à notre problématique <br />
et c'est la raison pour laquelle j'utiliserai leurs services.

L'objectif premier est de pouvoir, grâce à AWS, <u>louer de la puissance de calcul à la demande</u>. <br />
L'idée étant de pouvoir, quel que soit la charge de travail, <br />
obtenir suffisamment de puissance de calcul pour pouvoir traiter nos images, <br />
même si le volume de données venait à fortement augmenter.

De plus, la capacité d'utiliser cette puissance de calcul à la demande <br />
permet de diminuer drastiquement les coûts si l'on compare les coûts d'une location <br />
de serveur complet sur une durée fixe (1 mois, 1 année par exemple).


## 4.2 Choix de la solution technique : EMR

<u>Plusieurs solutions s'offre à nous</u> :
1. Solution **IAAS** (Infrastructure AS A Service)
 - Dans cette configuration **AWS** met à notre disposition des serveurs vierges <br />
   sur lequel nous avons un accès en administrateur, ils sont nommés **instance EC2**.<br />
   Pour faire simple, nous pouvons avec cette solution reproduire pratiquement <br />
   à l'identique la solution mis en œuvre en local sur notre machine.<br />
   <u>On installe nous-même l'intégralité des outils puis on exécute notre script</u> :
  - Installation de **Spark**, **Java** etc.
  - Installation de **Python** (via Anaconda par exemple)
  - Installation de **Jupyter Notebook**
  - Installation des **librairies complémentaires**
  - Il faudra bien évidement veiller à **implémenter les librairies
    nécessaires à toutes les machines (workers) du cluster**
  - <u>Avantages</u> :
      - Liberté totale de mise en œuvre de la solution
      - Facilité de mise en œuvre à partir d'un modèle qui s'exécute en local sur une machine Linux
  - <u>Inconvénients</u> :
      - Cronophage
          - Nécessité d'installer et de configurer toute la solution
      - Possible problèmes techniques à l'installation des outils (des problématiques qui <br />
        n'existaient pas en local sur notre machine peuvent apparaitre sur le serveur EC2)
      - Solution non pérenne dans le temps, il faudra veiller à la mise à jour des outils <br />
        et éventuellement devoir réinstaller Spark, Java etc.
2. Solution **PAAS** (Plateforme As A Service)
 - **AWS** fournit énormément de services différents, dans l'un de ceux-là <br />
   il existe une offre qui permet de louer des **instances EC2** <br />
   avec des applications préinstallées et configurées : il s'agit du **service EMR**.
 - **Spark** y sera déjà installé
 - Possibilité de demander l'installation de **Tensorflow** ainsi que **JupyterHub**
 - Possibilité d'indiquer des **packages complémentaires** à installer <br />
   à l'initialisation du serveur **sur l'ensemble des machines du cluster**.
 - <u>Avantages</u> :
     - Facilité de mise en œuvre
         - Il suffit de très peu de configuration pour obtenir <br />
           un environnement parfaitement fonctionnel
     - Rapidité de mise en œuvre
         - Une fois la première configuration réalisée, il est très facile <br />
           et très rapide de recréer des clusters à l'identique qui seront <br />
           disponibles presque instantanément (le temps d'instancier les <br />
           serveurs soit environ 15/20 minutes)
     - Solutions matérielless et logicielles optimisées par les ingénieurs d'AWS
         - On sait que les versions installées vont fonctionner <br />
           et que l'architecture proposée est optimisée
     - Stabilité de la solution
    - Solution évolutive
        Il est facile d’obtenir à chaque nouvelle instanciation une version à jour <br />
        de chaque package, en étant garanti de leur compatibilité avec le reste de l’environnement.
  - Plus sécurisé
	- Les éventuels patchs de sécurité seront automatiquement mis à jour <br />
      à chaque nouvelle instanciation du cluster EMR.
 - <u>Inconvénients</u> :
     - Peut-être un certain manque de liberté sur la version des packages disponibles ? <br />
       Même si je n'ai pas constaté ce problème.
   

Je retiens la solution **PAAS** en choisissant d'utiliser <br />
le service **EMR** d'Amazon Web Services.<br />
Je la trouve plus adaptée à notre problématique et permet <br />
une mise en œuvre qui soit à la fois plus rapide et <br />
plus efficace que la solution IAAS.

## 4.3 Choix de la solution de stockage des données : Amazon S3

<u>Amazon propose une solution très efficace pour la gestion du stockage des données</u> : **Amazon S3**. <br />
S3 pour Amazon Simple Storage Service.

Il pourrait être tentant de stocker nos données sur l'espace alloué par le serveur **EC2**, <br />
mais si nous ne prenons aucune mesure pour les sauvegarder ensuite sur un autre support, <br />
<u>les données seront perdues</u> lorsque le serveur sera résilié (on résilie le serveur lorsqu'on <br />
ne s'en sert pas pour des raisons de coût).<br />
De fait, si l'on décide d'utiliser l'espace disque du serveur EC2 il faudra imaginer <br />
une solution pour sauvegarder les données avant la résiliation du serveur.
De plus, nous serions exposés à certaines problématiques si nos données venaient à <br />
**saturer** l'espace disponible de nos serveurs (ralentissements, disfonctionnements).

<u>Utiliser **Amazon S3** permet de s'affranchir de toutes ces problématiques</u>. <br />
L'espace disque disponible est **illimité**, et il est **indépendant de nos serveurs EC2**. <br />
L'accès aux données est **très rapide** car nous restons dans l'environnement d'AWS <br />
et nous prenons soin de <u>choisir la même région pour nos serveurs **EC2** et **S3**</u>.

De plus, comme nous le verrons <u>il est possible d'accéder aux données sur **S3** <br />
    de la même manière que l'on **accède aux données sur un disque local**</u>.<br />
Nous utiliserons simplement un **PATH au format s3://...** .

## 4.4 Configuration de l'environnement de travail

La première étape est d'installer et de configurer [**AWS Cli**](https://aws.amazon.com/fr/cli/),<br />
il s'agit de l'**interface en ligne de commande d'AWS**.<br />
Elle nous permet d'**interagir avec les différents services d'AWS**, comme **S3** par exemple.

Pour pouvoir utiliser **AWS Cli**, il faut le configurer en créant préalablement <br />
un utilisateur à qui on donnera les autorisations dont nous aurons besoin.<br />
Dans ce projet il faut que l'utilisateur ait à minima un contrôle total sur le service S3.

<u>La gestion des utilisateurs et de leurs droits s'effectue via le service **AMI**</u> d'AWS.

Une fois l'utilisateur créé et ses autorisations configurées nous créons une **paire de clés** <br />
qui nous permettra de nous **connecter sans à avoir à devoir saisir systématiquement notre login/mot de passe**.<br />

Il faut également configurer l'**accès SSH** à nos futurs serveurs EC2. <br />
Ici aussi, via un système de clés qui nous dispense de devoir nous authentifier "à la main" à chaque connexion.

Toutes ses étapes de configuration sont parfaitement décrites <br />
dans le cours du projet: [Découvrez le cloud avec Amazon Web Services / Faites vos premiers pas sur AWS](https://openclassrooms.com/fr/courses/4810836-decouvrez-le-cloud-avec-amazon-web-services/7821712-faites-vos-premiers-pas-sur-aws)

## 4.5 Upload de nos données sur S3

Nos outils sont configurés. <br />
Il faut maintenant uploader nos données de travail sur Amazon S3.

Ici aussi les étapes sont décrites avec précision <br />
dans le cours [Découvrez le cloud avec Amazon Web Services / Stockez et accédez à des fichiers sur Amazon S3](https://openclassrooms.com/fr/courses/4810836-decouvrez-le-cloud-avec-amazon-web-services/7822690-stockez-et-accedez-a-des-fichiers-sur-amazon-s3)

Je décide de n'uploader que les données contenues dans le dossier **Test** du [jeu de données du projet](https://www.kaggle.com/moltean/fruits/download)


La première étape consiste à **créer un bucket sur S3** <br />
dans lequel nous uploaderons les données du projet:
- **aws s3 mb s3://p11-data-fruits-grgmdmn**

On vérifie que le bucket à bien été créé
- **aws s3 ls**
 - Si le nom du bucket s'affiche alors c'est qu'il a été correctement créé.

On copie ensuite le contenu du dossier "**Test**" <br />
dans un répertoire "**Test**" sur notre bucket "**p11-data-fruits-grgmdmn**":
1. On se place à l'intérieur du répertoire **Test**
2. **aws sync . s3://p11-data-fruits-grgmdmn/Test**

La commande **sync** est utile pour synchroniser deux répertoires.

<u>Nos données du projet sont maintenant disponibles sur Amazon S3</u>.

## 4.6 Configuration du serveur EMR

Une fois encore, le cours [Découvrez le cloud avec Amazon Web Services / Découvrez les services d'Amazon EC2](https://openclassrooms.com/fr/courses/4810836-decouvrez-le-cloud-avec-amazon-web-services/7822091-demarrez-votre-premiere-instance-ec2) <br /> détaille l'essentiel des étapes pour lancer un cluster avec **EMR**.

<u>Je détaillerai ici les étapes particulières qui nous permettent <br />
de configurer le serveur selon nos besoins</u> :

1. Cliquez sur Créer un cluster
![Créer un cluster](img/EMR_creer.png)
2. Cliquez sur Accéder aux options avancées
![Créer un cluster](img/EMR_options_avancees.png)

### 4.6.1 Étape 1 : Logiciels et étapes

#### 4.6.1.1 Configuration des logiciels

<u>Sélectionnez les packages dont nous aurons besoin comme dans la capture d'écran</u> :
1. Nous sélectionnons la dernière version d'**EMR**, soit la version **6.3.0** au moment où je rédige ce document
2. Nous cochons bien évidement **Hadoop** et **Spark** qui seront préinstallés dans leur version la plus récente
3. Nous aurons également besoin de **TensorFlow** pour importer notre modèle et réaliser le **transfert learning**
4. Nous travaillerons enfin avec un **notebook Jupyter** via l'application **JupyterHub**<br />
 - Comme nous le verrons dans un instant nous allons <u>paramétrer l'application afin que les notebooks</u>, <br />
   comme le reste de nos données de travail, <u>soient enregistrés directement sur S3</u>.
![Créer un cluster](img/EMR_configuration_logiciels.png)


#### 4.6.1.2 Modifier les paramètres du logiciel

<u>Paramétrez la persistance des notebooks créés et ouvert via JupyterHub</u> :
- On peut à cette étape effectuer des demandes de paramétrage particulières sur nos applications. <br />
  L'objectif est, comme pour le reste de nos données de travail, <br />
  d'éviter toutes les problématiques évoquées précédemment. <br />
  C'est l'objectif à cette étape, <u>nous allons enregistrer <br />
  et ouvrir les notebooks</u> non pas sur l'espace disque de  l'instance EC2 (comme <br />
  ce serait le cas dans la configuration par défaut de JupyterHub) mais <br />
  <u>directement sur **Amazon S3**</u>.
- <u>deux solutions sont possibles pour réaliser cela</u> :
 1. Créer un **fichier de configuration JSON** que l'on **upload sur S3** et on indique ensuite le chemin d’accès au fichier JSON
 2. Rentrez directement la configuration au format JSON

J'ai personnellement créé un fichier JSON lors de la création de ma première instance EMR, <br />
puis lorsqu'on décide de cloner notre serveur pour en recréer un facilement à l'identique, <br />
la configuration du fichier JSON se retrouve directement copié comme dans la capture ci-dessous.

<u>Voici le contenu de mon fichier JSON</u> :  [{"classification":"jupyter-s3-conf","properties":{"s3.persistence.bucket":"p11-data-fruits-grgmdmn","s3.persistence.enabled":"true"}}]
 Appuyez ensuite sur "**Suivant**"
![Modifier les paramètres du logiciel](img/EMR_parametres_logiciel.png)

### 4.6.2 Étape 2 : Matériel

A cette étape, laissez les choix par défaut. <br />
<u>L'important ici est la sélection de nos instances</u> :

1. je choisi les instances de type **M5** qui sont des **instances de type équilibrés**
2. je choisi le type **xlarge** qui est l'instance la **moins onéreuse disponible**
 [Plus d'informations sur les instances M5 Amazon EC2](https://aws.amazon.com/fr/ec2/instance-types/m5/)
3. Je sélectionne **1 instance Maître** (le driver) et **2 instances Principales** (les workeurs) <br />
   soit **un total de 3 instance EC2**.
![Choix du materiel](img/EMR_materiel.png)

### 4.6.3 Étape 3 : Paramètres de cluster généraux


#### 4.6.3.1 Options générales
<u>La première chose à faire est de donner un nom au cluster</u> :<br />
*J'ai également décoché "Protection de la résiliation" pour des raisons pratiques.*
    
![Nom du Cluster](img/EMR_nom_cluster.png)

#### 4.6.3.2 Actions d'amorçage

Nous allons à cette étape **choisir les packages manquants à installer** et qui <br />
nous serons utiles dans l'exécution de notre notebook.<br />
<u>L'avantage de réaliser cette étape maintenant est que les packages <br />
installés le seront sur l'ensemble des machines du cluster</u>.

Nous créons un fichier nommé "**bootstrap-emr.sh**" que nous <u>uploadons <br />
sur S3</u>(je l’installe à la racine de mon **bucket "p11-data-fruits-grgmdmn"**) et nous l'ajoutons <br />
comme indiqué dans la capture d'écran ci-dessous:
![Actions d'amorcage](img/EMR_amorcage.png)

Voici le contenu du fichier **bootstrap-emr.sh**<br />
Comme on peut le constater il s'agit simplement de commande "**pip install**" <br />
pour **installer les bibliothèques manquantes** comme réalisé en local.<br />
Une fois encore, <u>il est nécessaire de réaliser ces actions à cette étape</u> <br />
pour que <u>les packages soient installés sur l'ensemble des machines du cluster</u> <br />
et non pas uniquement sur le driver, comme cela serait le cas si nous exécutions <br />
ces commandes directement dans le notebook JupyterHub ou dans la console EMR (connecté au driver).
![Contenu du fichier bootstrap](img/EMR_bootstrap.png)

**setuptools** et **pip** sont mis à jour pour éviter une problématique <br />
avec l'installation du package **pyarrow**.<br />
**Pandas** a eu droit à une mise à jour majeur (1.3.0) il y a moins d'une semaine <br />
au moment de la rédaction de ce notebook, et la nouvelle version de **Pandas** <br />
nécessite une version plus récente de **Numpy** que la version installée par <br />
défaut (1.16.5) à l'initialisation des instances **EC2**. <u>Il ne semble pas <br />
possible d'imposer une autre version de Numpy que celle installé par <br />
défaut</u> même si on force l'installation d'une version récente de **Numpy** <br />
(en tout cas, ni simplement ni intuitivement).<br />
La mise à jour étant très récente <u>la version de **Numpy** n'est pas encore <br />
mise à jour sur **EC2**</u> mais on peut imaginer que ce sera le cas très rapidement <br />
et il ne sera plus nécessaire d'imposer une version spécifique de **Pandas**.<br />
En attendant, je demande <u>l'installation de l'avant dernière version de **Pandas (1.2.5)**</u>

On clique ensuite sur ***Suivant***

### 4.6.4 Étape 4 : Sécurité


#### 4.6.4.1 Options de sécurité

A cette étape nous sélectionnons la **paire de clés EC2** créé précédemment. <br />
Elle nous permettra de se connecter en **ssh** à nos **instances EC2** <br />
sans avoir à entrer nos login/mot de passe.<br />
On laisse les autres paramètres par défaut. <br />
Et enfin, on clique sur "***Créer un cluster***"

![EMR Sécurité](img/EMR_securite.png)

## 4.7 Instanciation du serveur

Il ne nous reste plus qu'à attendre que le serveur soit prêt. <br />
Cette étape peut prendre entre **15 et 20 minutes**.

<u>Plusieurs étapes s'enchaîne, on peut suivre l'avancé du statut du **cluster EMR**</u> :

![Instanciation étape 1](img/EMR_instanciation_01.png)
![Instanciation étape 2](img/EMR_instanciation_02.png)
![Instanciation étape 3](img/EMR_instanciation_03.png)

<u>Lorsque le statut affiche en vert: "**En attente**" cela signifie que l'instanciation <br />
s'est bien déroulée et que notre serveur est prêt à être utilisé</u>.

## 4.8 Création du tunnel SSH à l'instance EC2 (Maître)


### 4.8.1 Création des autorisations sur les connexions entrantes

<u>Nous souhaitons maintenant pouvoir accéder à nos applications</u> :
 - **JupyterHub** pour l'exécution de notre notebook
 - **Serveur d'historique Spark** pour le suivi de l'exécution <br />
   des tâches de notre script lorsqu'il sera lancé

Cependant, <u>ces applications ne sont accessibles que depuis le réseau local du driver</u>, <br />
et pour y accéder nous devons **créer un tunnel SSH vers le driver**.

Par défaut, ce driver se situe derrière un firewall qui bloque l'accès en SSH. <br />
<u>Pour ouvrir le port 22 qui correspond au port sur lequel écoute le serveur SSH, <br />
il faut modifier le **groupe de sécurité EC2 du driver**</u>.

*Il faudra que l'on se connecte en SSH au driver de notre cluster. <br />
Par défaut, ce driver se situe derrière un firewall qui bloque l'accès en SSH. <br />
Pour ouvrir le port 22 qui correspond au port sur lequel écoute le serveur SSH, <br />
il faut modifier le groupe de sécurité EC2 du driver. Sur la page de la console <br />
consacrée à EC2, dans l'onglet "Réseau et sécurité", cliquez sur "Groupes de sécurité". <br />
Vous allez devoir modifier le groupe de sécurité d’ElasticMapReduce-Master. <br />
Dans l'onglet "Entrant", ajoutez une règle SSH dont la source est "N'importe où" <br />
(ou "Mon IP" si vous disposez d'une adresse IP fixe).*

![Configuration autorisation ports entrants pour ssh](img/EMR_config_ssh_01.png)

<u>Une fois cette étape réalisée vous devriez avoir une configuration semblable à la mienne</u> :

![Configuration ssh terminée](img/EMR_config_ssh_02.png)

### 4.8.2 Création du tunnel ssh vers le Driver

On peut maintenant établir le **tunnel SSH** vers le **Driver**. <br />
Pour cela on récupère les informations de connexion fournis par Amazon <br />
depuis la page du service EMR / Cluster / onglet Récapitulatif en <br />
cliquant sur "**Activer la connexion Web**"

![Activer la connexion Web](img/EMR_tunnel_ssh_01.png)

<u>On récupère ensuite la commande fournis par Amazon pour **établir le tunnel SSH**</u> :

![Récupérer la commande pour établir le tunnel ssh](img/EMR_tunnel_ssh_02.png)

<u>Dans mon cas, la commande ne fonctionne pas tel</u> quel et j'ai du **l'adapter à ma configuration**. <br />
La **clé ssh** se situe dans un dossier "**.ssh**" elle-même située dans <br />
mon **répertoire personnel** dont le symbole est, sous Linux, identifié par un tilde "**~**".

<u>Finalement, j'utilise la commande suivante dans un terminal pour établir <br />
    mon tunnel ssh (seul l'URL change d'une instance à une autre)</u> : <br />
"**ssh -i ~/.ssh/p11-ec2.pem -D 5555 hadoop@ec2-35-180-91-39.eu-west-3.compute.amazonaws.com**"

<u>On inscrit "**yes**" pour valider la connexion et si <br />
    la connexion est établit on obtient le résultat suivant</u> :

![Création du tunnel SSH](img/EMR_connexion_ssh_01.png)

Nous avons **correctement établi le tunnel ssh avec le driver** sur le port "5555".

### 4.8.3 Configuration de FoxyProxy

Une dernière étape est nécessaire pour accéder à nos applications, <br />
en demandant à notre navigateur d'emprunter le tunnel ssh.<br />
J'utilise pour cela **FoxyProxy**.

Sinon, ouvrez la configuration de **FoxyProxy** et <u>cliquez sur **Ajouter**</u> en haut à gauche <br />
puis renseigner les éléments comme dans la capture ci-dessous :

![Configuration FoxyProxy Etape 1](img/EMR_foxyproxy_config_01.png)

<u>On obtient le résultat ci-dessous</u> :

![Configuration FoxyProxy Etape 2](img/EMR_foxyproxy_config_02.png)

### 4.8.4 Accès aux applications du serveur EMR via le tunnel ssh


<u>Avant d'établir notre **tunnel ssh** nous avions ça</u> :

![avant tunnel ssh](img/EMR_tunnel_ssh_avant.png)

<u>On active le **tunnel ssh** comme vu précédemment puis on demande <br />
à notre navigateur de l'utiliser avec **FoxyProxy**</u> :

![FoxyProxy activation](img/EMR_foxyproxy_activation.png)

<u>On peut maintenant s'apercevoir que plusieurs applications nous sont accessibles</u> :

![avant tunnel ssh](img/EMR_tunnel_ssh_apres.png)

## 4.9 Connexion au notebook JupyterHub

Pour se connecter à **JupyterHub** en vue d'exécuter notre **notebook**, <br />
il faut commencer par <u>cliquer sur l'application **JupyterHub**</u> apparu <br />
depuis que nous avons configuré le **tunnel ssh** et **foxyproxy** sur <br />
notre navigateur (actualisez la page si ce n’est pas le cas).

![Démarrage de JupyterHub](img/EMR_jupyterhub_connexion_01.png)

On passe les éventuels avertissements de sécurité puis <br />
nous arrivons sur une page de connexion.
    
<u>On se connecte avec les informations par défaut</u> :
 - <u>login</u>: **jovyan**
 - <u>password</u>: **jupyter**

![Connexion à JupyterHub](img/EMR_jupyterhub_connexion_02.png)

Nous arrivons ensuite dans un dossier vierge de notebook.<br />
Il suffit d'en créer un en cliquant sur "**New**" en haut à droite.

![Liste et création des notebook](img/EMR_jupyterhub_creer_notebooks.png)

Il est également possible d'en <u>uploader un directement dans notre **bucket S3**</u>.

Grace à la <u>**persistance** paramétrée à l'instanciation du cluster <br />
nous sommes actuellement dans l'arborescence de notre **bucket S3**</u>

![Notebook stockés sur S3](img/EMR_jupyterhub_S3.png)

Je décide d'**importer un notebook déjà rédigé en local directement <br />
sur S3** et je l'ouvre depuis **l'interface JupyterHub**.



## 4.10 Exécution du code

Je décide d'exécuter cette partie du code depuis **JupyterHub hébergé sur notre cluster EMR**.<br />
Pour ne pas alourdir inutilement les explications du **notebook**, je ne réexpliquerai pas les étapes communes <br />
que nous avons déjà vues dans la première partie où l'on a exécuté le code localement sur notre machine virtuelle Ubuntu.

<u>Avant de commencer</u>, il faut s'assurer d'utiliser le **kernel pyspark**.

**En utilisant ce kernel, une session spark est créé à l'exécution de la première cellule**. <br />
Il n'est donc **plus nécessaire d'exécuter le code "spark = (SparkSession ..."** comme lors <br />
de l'exécution de notre notebook en local sur notre VM Ubuntu.

### 4.10.1 Démarrage de la session Spark

In [1]:
# L'exécution de cette cellule démarre l'application Spark

Starting Spark application


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

<u>Affichage des informations sur la session en cours et liens vers Spark UI</u> :

In [2]:
%%info

### 4.10.2 Installation des packages

Les packages nécessaires ont été installé via l'étape de **bootstrap** à l'instanciation du serveur.

### 4.10.3 Import des librairies

In [3]:
import pandas as pd
import numpy as np
import io
import os

import tensorflow as tf
from PIL import Image
from tensorflow.keras.applications.mobilenet_v2 import MobileNetV2, preprocess_input
from tensorflow.keras.preprocessing.image import img_to_array
from tensorflow.keras import Model
from pyspark.sql.functions import col, pandas_udf, PandasUDFType, element_at, split, udf
from pyspark.ml.feature import PCA
from pyspark.ml.linalg import Vectors, VectorUDT  # ← VectorUDT est dans pyspark.ml.linalg

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

### 4.10.4 Définition des PATH pour charger les images et enregistrer les résultats

Nous accédons directement à nos **données sur S3** comme si elles étaient **stockées localement**.

In [4]:
PATH = 's3://p11-data-fruits-grgmdmn'
PATH_Data = PATH+'/test-multiple-fruits'
PATH_Result = PATH+'/Results'
print('PATH:        '+\
      PATH+'\nPATH_Data:   '+\
      PATH_Data+'\nPATH_Result: '+PATH_Result)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

PATH:        s3://p11-data-fruits-grgmdmn
PATH_Data:   s3://p11-data-fruits-grgmdmn/test-multiple-fruits
PATH_Result: s3://p11-data-fruits-grgmdmn/Results

### 4.10.5 Traitement des données

#### 4.10.5.1 Chargement des données

In [5]:
images = spark.read.format("binaryFile") \
  .option("pathGlobFilter", "*.jpg") \
  .option("recursiveFileLookup", "true") \
  .load(PATH_Data)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [6]:
images.show(5)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

+--------------------+-------------------+-------+--------------------+
|                path|   modificationTime| length|             content|
+--------------------+-------------------+-------+--------------------+
|s3://p11-data-fru...|2025-09-22 16:37:51|6457796|[FF D8 FF E1 03 F...|
|s3://p11-data-fru...|2025-09-22 16:38:03|5704339|[FF D8 FF E1 03 F...|
|s3://p11-data-fru...|2025-09-22 16:38:06|5612220|[FF D8 FF E1 03 4...|
|s3://p11-data-fru...|2025-09-22 16:38:04|5601970|[FF D8 FF E1 03 F...|
|s3://p11-data-fru...|2025-09-22 16:37:59|5443640|[FF D8 FF E1 7C 5...|
+--------------------+-------------------+-------+--------------------+
only showing top 5 rows

<u>Je ne conserve que le **path** de l'image et j'ajoute <br />
    une colonne contenant les **labels** de chaque image</u> :

In [7]:
images = images.withColumn('label', element_at(split(images['path'], '/'),-2))
print(images.printSchema())
print(images.select('path','label').show(5,False))

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

root
 |-- path: string (nullable = true)
 |-- modificationTime: timestamp (nullable = true)
 |-- length: long (nullable = true)
 |-- content: binary (nullable = true)
 |-- label: string (nullable = true)

None
+-------------------------------------------------------------------+--------------------+
|path                                                               |label               |
+-------------------------------------------------------------------+--------------------+
|s3://p11-data-fruits-grgmdmn/test-multiple-fruits/cherries.jpg     |test-multiple-fruits|
|s3://p11-data-fruits-grgmdmn/test-multiple-fruits/strawberries.jpg |test-multiple-fruits|
|s3://p11-data-fruits-grgmdmn/test-multiple-fruits/apples4.jpg      |test-multiple-fruits|
|s3://p11-data-fruits-grgmdmn/test-multiple-fruits/strawberries5.jpg|test-multiple-fruits|
|s3://p11-data-fruits-grgmdmn/test-multiple-fruits/plums_pears.jpg  |test-multiple-fruits|
+-------------------------------------------------------------

#### 4.10.5.2 Préparation du modèle

In [8]:
model = MobileNetV2(weights='imagenet',
                    include_top=True,
                    input_shape=(224, 224, 3))

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

14540800/14536120 [==============================] - 1s 0us/step

In [9]:
new_model = Model(inputs=model.input,
                  outputs=model.layers[-2].output)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [10]:
brodcast_weights = sc.broadcast(new_model.get_weights())

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [11]:
new_model.summary()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 224, 224, 3) 0                                            
__________________________________________________________________________________________________
Conv1 (Conv2D)                  (None, 112, 112, 32) 864         input_1[0][0]                    
__________________________________________________________________________________________________
bn_Conv1 (BatchNormalization)   (None, 112, 112, 32) 128         Conv1[0][0]                      
__________________________________________________________________________________________________
Conv1_relu (ReLU)               (None, 112, 112, 32) 0           bn_Conv1[0][0]                   
______________________________________________________________________________________________

In [12]:
def model_fn():
    """
    Returns a MobileNetV2 model with top layer removed
    and broadcasted pretrained weights.
    """
    model = MobileNetV2(weights=None, # There is no need to download pre-trained imagenet weights, as we already have broadcast_weights.
                        include_top=True,
                        input_shape=(224, 224, 3))
    for layer in model.layers:
        layer.trainable = False
    new_model = Model(inputs=model.input,
                  outputs=model.layers[-2].output)
    new_model.set_weights(brodcast_weights.value)
    return new_model

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

#### 4.10.5.3 Définition du processus de chargement des images <br/> et application de leur featurisation à travers l'utilisation de pandas UDF

In [13]:
def preprocess(content):
    """
    Preprocesses raw image bytes for prediction.
    """
    img = Image.open(io.BytesIO(content)).resize([224, 224])
    arr = img_to_array(img)
    return preprocess_input(arr)

def featurize_series(model, content_series):
    """
    Featurize a pd.Series of raw images using the input model.
    :return: a pd.Series of image features
    """
    input = np.stack(content_series.map(preprocess))
    preds = model.predict(input)
    # For some layers, output features will be multi-dimensional tensors.
    # We flatten the feature tensors to vectors for easier storage in Spark DataFrames.
    output = [p.flatten() for p in preds]
    return pd.Series(output)

@pandas_udf('array<float>', PandasUDFType.SCALAR_ITER)
def featurize_udf(content_series_iter):
    '''
    This method is a Scalar Iterator pandas UDF wrapping our featurization function.
    The decorator specifies that this returns a Spark DataFrame column of type ArrayType(FloatType).

    :param content_series_iter: This argument is an iterator over batches of data, where each batch
                              is a pandas Series of image data.
    '''
    # With Scalar Iterator pandas UDFs, we can load the model once and then re-use it
    # for multiple data batches.  This amortizes the overhead of loading big models.
    model = model_fn()
    for content_series in content_series_iter:
        yield featurize_series(model, content_series)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

/usr/lib/spark/python/lib/pyspark.zip/pyspark/sql/pandas/functions.py:392: UserWarning: In Python 3.6+ and Spark 3.0+, it is preferred to specify type hints for pandas UDF instead of specifying pandas UDF type which will be deprecated in the future releases. See SPARK-28264 for more details.

#### 4.10.5.4 Exécutions des actions d'extractions de features

In [14]:
# spark.conf.set("spark.sql.execution.arrow.maxRecordsPerBatch", "1024")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [18]:
features_df = images.repartition(20).select(col("path"),
                                            col("label"),
                                            featurize_udf("content").alias("features")
                                           )

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [19]:
print(PATH_Result)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

s3://p11-data-fruits-grgmdmn/Results

In [20]:
# On cherche à réduire la dimensionalité afin de gagner en espace de stockage ET profiter d'un temps d'entraînement réduit
# features_df.write.mode("overwrite").parquet(PATH_Result)

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

#### 4.10.5.5 Exécution d'une PCA distribuée

#### Limitation

Il est important de noter que la présence d'un faible échantillon (exemple minimal) va influer sur la variance captée par chacune des composantes principales. 100 composantes principales pour un échantillon de test de 102 images va forcément mathématiquement capter presque 100% de la variance. Cela sera bien moins vrai pour un échantillon de plus de 20 000 images. La variable `k_components` sera donc probablement à ajuster, en particulier dans l'environnement cloud.

In [21]:
k_components = 70

# Convertir les features en format ML de Spark
def to_vector(features):
    return Vectors.dense(features)

vector_udf = udf(to_vector, VectorUDT())

# Préparer les données pour la PCA
features_vector_df = features_df.withColumn("features_vector", vector_udf("features"))

# Appliquer la PCA pour réduire de 1280 à k_components dimensions
pca = PCA(k=k_components, inputCol="features_vector", outputCol="pca_features")
pca_model = pca.fit(features_vector_df)
pca_result = pca_model.transform(features_vector_df)

# Afficher la variance expliquée par les composantes principales
explained_variance = pca_model.explainedVariance
total_variance = sum(explained_variance)
print(f"Variance expliquée par les {k_components} premières composantes : {total_variance:.4f} ({total_variance*100:.2f}%)")
print(f"Réduction de dimensionnalité : 1280 → {k_components} dimensions")

# Sélectionner les colonnes finales
final_df = pca_result.select("path", "label", "pca_features")

# Sauvegarder les résultats avec PCA sur S3
PATH_Result_PCA = PATH + '/Results_PCA'
final_df.write.mode("overwrite").parquet(PATH_Result_PCA)

print(f"Résultats sauvés sur S3 : {PATH_Result_PCA}")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Variance expliquée par les 70 premières composantes : 0.9994 (99.94%)
Réduction de dimensionnalité : 1280 → 70 dimensions
Résultats sauvés sur S3 : s3://p11-data-fruits-grgmdmn/Results_PCA

### 4.10.6 Chargement des données enregistrées et validation du résultat

In [28]:
df_pca = pd.read_parquet(PATH_Result_PCA, engine='pyarrow')

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [29]:
df_pca.head()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

                                                path  ...                                       pca_features
0  s3://p11-data-fruits-grgmdmn/test-multiple-fru...  ...  {'type': 1, 'size': None, 'indices': None, 'va...
1  s3://p11-data-fruits-grgmdmn/test-multiple-fru...  ...  {'type': 1, 'size': None, 'indices': None, 'va...
2  s3://p11-data-fruits-grgmdmn/test-multiple-fru...  ...  {'type': 1, 'size': None, 'indices': None, 'va...
3  s3://p11-data-fruits-grgmdmn/test-multiple-fru...  ...  {'type': 1, 'size': None, 'indices': None, 'va...
4  s3://p11-data-fruits-grgmdmn/test-multiple-fru...  ...  {'type': 1, 'size': None, 'indices': None, 'va...

[5 rows x 3 columns]

In [30]:
# Explorer la structure du vecteur PCA
pca_vector = df_pca.loc[0,'pca_features']
print(f"Type: {type(pca_vector)}")
print(f"Clés disponibles: {pca_vector.keys()}")

if 'values' in pca_vector:
    values = pca_vector['values']
    print(f"Nombre de dimensions PCA: {len(values)}")
    print(f"Premières valeurs: {values[:5]}")

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Type: <class 'dict'>
Clés disponibles: dict_keys(['type', 'size', 'indices', 'values'])
Nombre de dimensions PCA: 70
Premières valeurs: [ 1.74324346e+01  3.79323925e-03  7.61737608e-01  1.10952834e+00
 -2.34927779e+00]

In [31]:
df_pca.shape

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

(102, 3)

In [32]:
df_pca["pca_features"].loc[0]['values'].shape

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

(70,)

<u>On peut également constater la présence des fichiers <br />
    au format "**parquet**" sur le **serveur S3**</u> :

![Affichage des résultats sur S3](img/S3_Results.png)

## 4.11 Suivi de l'avancement des tâches avec le Serveur d'Historique Spark

Il est possible de voir l'avancement des tâches en cours <br />
avec le **serveur d'historique Spark**.

![Accès au serveur d'historique spark](img/EMR_serveur_historique_spark_acces.png)

**Il est également possible de revenir et d'étudier les tâches <br />
qui ont été réalisé, afin de debugger, optimiser les futurs <br />
tâches à réaliser.**

<u>Lorsque la commande "**features_df.write.mode("overwrite").parquet(PATH_Result)**" <br />
était en cours, nous pouvions observer son état d'avancement</u> :

![Progression execution script](img/EMR_jupyterhub_avancement.png)

<u>Le **serveur d'historique Spark** nous permet une vision beaucoup plus précise <br />
de l'exécution des différentes tâche sur les différentes machines du cluster</u> :

![Suivi des tâches spark](img/EMR_SHSpark_01.png)

On peut également constater que notre cluster de calcul a mis <br />
un tout petit peu **moins de 8 minutes** pour traiter les **22 688 images**.

![Temps de traitement](img/EMR_SHSpark_02.png)


## 4.12 Résiliation de l'instance EMR

Notre travail est maintenant terminé. <br />
Le cluster de machines EMR est **facturé à la demande**, <br />
et nous continuons d'être facturé même lorsque <br />
les machines sont au repos.<br />
Pour **optimiser la facturation**, il nous faut <br />
maintenant **résilier le cluster**.

<u>Je réalise cette commande depuis l'interface AWS</u> :

1. Commencez par **désactiver le tunnel ssh dans FoxyProxy** pour éviter des problèmes de **timeout**.
![Désactivation de FoxyProxy](img/EMR_foxyproxy_desactivation.png)
2. Cliquez sur "**Résilier**"
![Cliquez sur Résilier](img/EMR_resiliation_01.png)
3. Confirmez la résiliation
![Confirmez la résiliation](img/EMR_resiliation_02.png)
4. La résiliation prend environ **1 minute**
![Résiliation en cours](img/EMR_resiliation_03.png)
5. La résiliation est effectuée
![Résiliation terminée](img/EMR_resiliation_04.png)

## 4.13 Cloner le serveur EMR (si besoin)

Si nous devons de nouveau exécuter notre notebook dans les mêmes conditions, <br />
il nous suffit de **cloner notre cluster** et ainsi en obtenir une copie fonctionnelle <br />
sous 15/20 minutes, le temps de son instanciation.

<u>Pour cela deux solutions</u> :
1. <u>Depuis l'interface AWS</u> :
 1. Cliquez sur "**Cloner**"
   ![Cloner un cluster](img/EMR_cloner_01.png)
 2. Dans notre cas nous ne souhaitons pas inclure d'étapes
   ![Ne pas inclure d'étapes](img/EMR_cloner_02.png)
 3. La configuration du cluster est recréée à l’identique. <br />
    On peut revenir sur les différentes étapes si on souhaite apporter des modifications<br />
    Quand tout est prêt, cliquez sur "**Créer un cluster**"
  ![Vérification/Modification/Créer un cluster](img/EMR_cloner_03.png)
2. <u>En ligne de commande</u> (avec AWS CLI d'installé et de configuré et en s'assurant <br />
   de s'attribuer les droits nécessaires sur le compte AMI utilisé)
 1. Cliquez sur "**Exporter AWS CLI**"
 ![Exporter AWS CLI](img/EMR_cloner_cli_01.png)
 2. Copier/Coller la commande **depuis un terminal**
 ![Copier Coller Commande](img/EMR_cloner_cli_02.png)


## 4.14 Arborescence du serveur S3 à la fin du projet

<u>Pour information, voici **l'arborescence complète de mon bucket S3 p11-data-fruits-grgmdmn** à la fin du projet</u> : <br />
*Par soucis de lisibilité, je ne liste pas les 131 sous dossiers du répertoire "Test"*

2025-09-29 15:28:30          0 Results_PCA/_SUCCESS

2025-09-29 15:28:19       5318 Results_PCA/part-00000-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:19       4713 Results_PCA/part-00001-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:19       4663 Results_PCA/part-00002-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:19       5212 Results_PCA/part-00003-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:21       5224 Results_PCA/part-00004-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:21       4634 Results_PCA/part-00005-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:21       5167 Results_PCA/part-00006-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:21       5174 Results_PCA/part-00007-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:24       5774 Results_PCA/part-00008-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:24       6326 Results_PCA/part-00009-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:24       5773 Results_PCA/part-00010-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:24       5788 Results_PCA/part-00011-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:26       5156 Results_PCA/part-00012-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:27       5224 Results_PCA/part-00013-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:27       5292 Results_PCA/part-00014-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:27       5261 Results_PCA/part-00015-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:29       5197 Results_PCA/part-00016-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:29       5162 Results_PCA/part-00017-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:29       5179 Results_PCA/part-00018-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet

2025-09-29 15:28:30       5170 Results_PCA/part-00019-22bc6605-c69c-4bae-ad9c-d0394efb4a27-c000.snappy.parquet  

2025-09-25 16:34:47        219 bootstrap-emr.sh

2025-09-22 16:40:56        135 jupyter-s3-conf.json

2025-09-25 16:17:16          0 jupyter/jovyan/.s3keep

2025-09-25 16:41:43     466644 jupyter/jovyan/P11_Notebook_Linux_EMR_PySpark_V1.1.ipynb

2025-09-22 16:37:50    1402274 test-multiple-fruits/Bananas(lady_finger)1.jpg

2025-09-22 16:37:48    1565087 test-multiple-fruits/Bananas(lady_finger)2.jpg

2025-09-22 16:37:49    1743367 test-multiple-fruits/Bananas(lady_finger)3.jpg

2025-09-22 16:37:49    1854256 test-multiple-fruits/Bananas(lady_finger)4.jpg

2025-09-22 16:37:50    1478502 test-multiple-fruits/Bananas(lady_finger)5.jpg

2025-09-22 16:37:50    2671709 test-multiple-fruits/apple.jpg

2025-09-22 16:37:50    2282152 test-multiple-fruits/apple_apricot_peach_peach(flat)_pomegranate_pear_plum.jpg       

2025-09-22 16:37:50    1890424 test-multiple-fruits/apple_apricot_peach_peach(flat)_pomegranate_pear_plum_2.jpg     

2025-09-22 16:37:50    2309053 test-multiple-fruits/apple_apricot_peach_peach(flat)_pomegranate_pear_plum_3.jpg

2025-09-22 16:37:50    3704053 test-multiple-fruits/apple_grape.jpg

2025-09-22 16:37:50    2980435 test-multiple-fruits/apple_grape2.jpg

2025-09-22 16:37:50    4640120 test-multiple-fruits/apple_pear.jpg

2025-09-22 16:37:49    4663157 test-multiple-fruits/apples1.jpg

2025-09-22 16:37:50    5248448 test-multiple-fruits/apples2.jpg

2025-09-22 16:38:06    5138167 test-multiple-fruits/apples3.jpg

2025-09-22 16:38:06    5612220 test-multiple-fruits/apples4.jpg

2025-09-22 16:38:06    4144168 test-multiple-fruits/apples_peaches1.jpg

2025-09-22 16:38:06    3983349 test-multiple-fruits/apples_peaches2.jpg

2025-09-22 16:37:51    4107025 test-multiple-fruits/apples_pears.jpg

2025-09-22 16:37:51     560855 test-multiple-fruits/cactus fruits.jpg

2025-09-22 16:37:51    3642803 test-multiple-fruits/cherries(rainier)1.jpg

2025-09-22 16:37:51    2093916 test-multiple-fruits/cherries(rainier)2.jpg

2025-09-22 16:37:51    6457796 test-multiple-fruits/cherries.jpg

2025-09-22 16:37:51    3840277 test-multiple-fruits/cherries2.jpg

2025-09-22 16:37:52    3891655 test-multiple-fruits/cherries3.jpg

2025-09-22 16:37:52    3027964 test-multiple-fruits/cherries4.jpg

2025-09-22 16:37:52    3512478 test-multiple-fruits/cherries5.jpg

2025-09-22 16:37:52    4074333 test-multiple-fruits/cherries6.jpg

2025-09-22 16:37:52    2909435 test-multiple-fruits/cherries7.jpg

2025-09-22 16:37:53    2933142 test-multiple-fruits/cherries8.jpg

2025-09-22 16:37:53    3886519 test-multiple-fruits/cherries_strawberries1.jpg

2025-09-22 16:37:53    4649575 test-multiple-fruits/cherries_strawberries2.jpg

2025-09-22 16:37:53    2085744 test-multiple-fruits/cherries_wax1.jpg

2025-09-22 16:37:54    5045853 test-multiple-fruits/cherries_wax2.jpg

2025-09-22 16:37:54    5238999 test-multiple-fruits/cherries_wax3.jpg

2025-09-22 16:37:54    5091251 test-multiple-fruits/cherries_wax4.jpg

2025-09-22 16:37:54    3915070 test-multiple-fruits/cherries_wax5.jpg

2025-09-22 16:37:54    4239368 test-multiple-fruits/cherries_wax6.jpg

2025-09-22 16:37:54    2196420 test-multiple-fruits/cherries_wax7.jpg

2025-09-22 16:37:54    4884999 test-multiple-fruits/cherry2.jpg

2025-09-22 16:37:55    2909435 test-multiple-fruits/cherry3.jpg

2025-09-22 16:37:55    3886519 test-multiple-fruits/cherry_strawberries.jpg

2025-09-22 16:37:55    4649575 test-multiple-fruits/cherry_strawberries2.jpg

2025-09-22 16:37:55    2700441 test-multiple-fruits/chestnut1.jpg

2025-09-22 16:37:55    2233752 test-multiple-fruits/chestnut3.jpg

2025-09-22 16:37:56    4305631 test-multiple-fruits/cocos.jpg

2025-09-22 16:37:56    3630892 test-multiple-fruits/cocos_kiwi_orange_dates_salak_plum_tamarilo_maracuja.jpg        

2025-09-22 16:37:56    3580646 test-multiple-fruits/cocos_kiwi_orange_dates_salak_plum_tamarilo_maracuja2.jpg       

2025-09-22 16:37:56    2759163 test-multiple-fruits/cocos_kiwi_orange_dates_salak_plum_tamarilo_maracuja3.jpg       

2025-09-22 16:37:56    4300685 test-multiple-fruits/dates.jpg

2025-09-22 16:37:56    3977020 test-multiple-fruits/grape_pear_mandarine.jpg

2025-09-22 16:37:56    4687081 test-multiple-fruits/grape_pear_mandarine2.jpg

2025-09-22 16:37:57    3430802 test-multiple-fruits/huckleberry.jpg

2025-09-22 16:37:57    2768816 test-multiple-fruits/kachi.jpg

2025-09-22 16:37:57    4771421 test-multiple-fruits/kiwi.jpg

2025-09-22 16:37:57    2970459 test-multiple-fruits/limes.jpg

2025-09-22 16:37:58    1690761 test-multiple-fruits/lychee.jpg

2025-09-22 16:37:58    1781700 test-multiple-fruits/mango.jpg

2025-09-22 16:37:58     957490 test-multiple-fruits/mangos1.jpg

2025-09-22 16:37:58    3997008 test-multiple-fruits/mangos2.jpg

2025-09-22 16:37:58     744345 test-multiple-fruits/mangos3.jpg

2025-09-22 16:37:58    1781124 test-multiple-fruits/mangos_bananas(lady_finger).jpg

2025-09-22 16:37:58    4683973 test-multiple-fruits/peaches.jpg

2025-09-22 16:37:58    4340231 test-multiple-fruits/pear.jpg

2025-09-22 16:37:59    4476078 test-multiple-fruits/pears1.jpg

2025-09-22 16:37:59    5044001 test-multiple-fruits/pears2.jpg

2025-09-22 16:37:59    2196773 test-multiple-fruits/physalis1.jpg

2025-09-22 16:37:59    4747944 test-multiple-fruits/physalis2.jpg

2025-09-22 16:37:59    5443640 test-multiple-fruits/plums_pears.jpg

2025-09-22 16:37:59    3532179 test-multiple-fruits/pomegranate.jpg

2025-09-22 16:38:00    3557576 test-multiple-fruits/pomegranate2.jpg

2025-09-22 16:38:00    3845675 test-multiple-fruits/raspberries1.jpg

2025-09-22 16:38:00    3708562 test-multiple-fruits/raspberries10.jpg

2025-09-22 16:38:00    4272462 test-multiple-fruits/raspberries11.jpg

2025-09-22 16:38:00    2939020 test-multiple-fruits/raspberries12.jpg

2025-09-22 16:38:01    4051706 test-multiple-fruits/raspberries2.jpg

2025-09-22 16:38:01    3839989 test-multiple-fruits/raspberries3.jpg

2025-09-22 16:38:01    4013902 test-multiple-fruits/raspberries4.jpg

2025-09-22 16:38:01    3642003 test-multiple-fruits/raspberries5.jpg

2025-09-22 16:38:01    5307375 test-multiple-fruits/raspberries6.jpg

2025-09-22 16:38:01    3390907 test-multiple-fruits/raspberries7.jpg

2025-09-22 16:38:01    3541222 test-multiple-fruits/raspberries8.jpg

2025-09-22 16:38:02    3378273 test-multiple-fruits/raspberries9.jpg

2025-09-22 16:38:02    4289161 test-multiple-fruits/raspberry.jpg

2025-09-22 16:38:02    3839989 test-multiple-fruits/raspberry2.jpg

2025-09-22 16:38:03    4013902 test-multiple-fruits/raspberry3.jpg

2025-09-22 16:38:03    5307375 test-multiple-fruits/raspberry4.jpg

2025-09-22 16:38:03    3541222 test-multiple-fruits/raspberry5.jpg

2025-09-22 16:38:03    5325447 test-multiple-fruits/raspberry_apple.jpg

2025-09-22 16:38:03    3317624 test-multiple-fruits/salak.jpg

2025-09-22 16:38:03    3638398 test-multiple-fruits/salak2.jpg

2025-09-22 16:38:03    5704339 test-multiple-fruits/strawberries.jpg

2025-09-22 16:38:04    4990351 test-multiple-fruits/strawberries2.jpg

2025-09-22 16:38:04    3210308 test-multiple-fruits/strawberries3.jpg

2025-09-22 16:38:04    4453982 test-multiple-fruits/strawberries4.jpg

2025-09-22 16:38:04    5601970 test-multiple-fruits/strawberries5.jpg

2025-09-22 16:38:04    2918414 test-multiple-fruits/tangelo.jpg

2025-09-22 16:38:05    4358384 test-multiple-fruits/tomatoes1.jpg

2025-09-22 16:38:05    4489489 test-multiple-fruits/tomatoes2.jpg

2025-09-22 16:38:05    4588272 test-multiple-fruits/tomatoes3.jpg

2025-09-22 16:38:06    4211396 test-multiple-fruits/tomatoes4.jpg

2025-09-22 16:38:06    5146765 test-multiple-fruits/tomatoes5.jpg





# 5. Conclusion

Nous avons réalisé ce projet **en deux temps** en tenant <br />
compte des contraintes qui nous ont été imposées.

Nous avons **dans un premier temps développé notre solution en local** <br />
sur une machine virtuelle dans un environnement Linux Ubuntu.

La <u>première phase</u> a consisté à **installer l'environnement de travail Spark**. <br />
**Spark** a un paramètre qui nous permet de travaillé en local et nous permet <br />
ainsi de **simuler du calcul partagé** en considérant <br />
**chaque cœur d'un processeur comme un worker indépendant**.<br />
Nous avons travaillé sur un plus **petit jeu de donnée**, l'idée était <br />
simplement de **valider le bon fonctionnement de la solution**.

Nous avons fait le choix de réaliser du **transfert learning** <br />
à partir du model **MobileNetV2**.<br />
Ce modèle a été retenu pour sa **légèreté** et sa **rapidité d'exécution** <br />
ainsi que pour la **faible dimension de son vecteur en sortie**.

Les résultats ont été enregistrés sur disque en plusieurs <br />
partitions au format "**parquet**".

<u>**La solution a parfaitement fonctionné en mode local**</u>.

La <u>deuxième phase</u> a consisté à créer un **réel cluster de calculs**. <br />
L'objectif était de pouvoir **anticiper une future augmentation de la charge de travail**.

Le meilleur choix retenu a été l'utilisation du prestataire de services **Amazon Web Services** <br />
qui nous permet de **louer à la demande de la puissance de calculs**, <br />
pour un **coût tout à fait acceptable**.<br />
Ce service se nomme **EC2** et se classe parmi les offres **Infrastructure As A Service** (IAAS).

Nous sommes allez plus loin en utilisant un service de plus <br />
haut niveau (**Plateforme As A Service** PAAS)<br />
en utilisant le service **EMR** qui nous permet d'un seul coup <br />
d'**instancier plusieurs serveur (un cluster)** sur lesquels <br />
nous avons pu demander l'installation et la configuration de plusieurs<br />
programmes et librairies nécessaires à notre projet comme **Spark**, <br />
**Hadoop**, **JupyterHub** ainsi que la librairie **TensorFlow**.

En plus d'être plus **rapide et efficace à mettre en place**, nous avons <br />
la **certitude du bon fonctionnement de la solution**, celle-ci ayant été <br />
préalablement validé par les ingénieurs d'Amazon.

Nous avons également pu installer, sans difficulté, **les packages <br />
nécessaires sur l'ensembles des machines du cluster**.

Enfin, avec très peu de modification, et plus simplement encore, <br />
nous avons pu **exécuter notre notebook comme nous l'avions fait localement**.<br />
Nous avons cette fois-ci exécuté le traitement sur **l'ensemble des images de notre dossier "Test"**.

Nous avons opté pour le service **Amazon S3** pour **stocker les données de notre projet**. <br />
S3 offre, pour un faible coût, toutes les conditions dont nous avons besoin pour stocker <br />
et exploiter de manière efficace nos données.<br />
L'espace alloué est potentiellement **illimité**, mais les coûts seront fonction de l'espace utilisé.

Il nous sera **facile de faire face à une monté de la charge de travail** en **redimensionnant** <br />
simplement notre cluster de machines (horizontalement et/ou verticalement au besoin), <br />
les coûts augmenteront en conséquence mais resteront nettement inférieurs aux coûts engendrés <br />
par l'achat de matériels ou par la location de serveurs dédiés.